### DISCUSSION TITLE
Docker-in-docker workflow (dind)

### DISCUSSION THREAD

**Discussion by alexey-igrychev** [werf/werf#7180] [General]
Currently, werf builds images only with **local Docker daemon**. Thus, the docker-in-docker approach is possible but socket binding (`/var/run/docker.sock`) is required.

During the build process, werf uses binding for:
* git archives and patches;
* Ansible playbooks and shell scripts;
* user mounts;
* imports.

Therefore, the container should run in privileged mode and the following host directories should be there:
- `tmp-dir`: dir to store tmp files and dirs (default $WERF_TMP_DIR or system tmp dir);
- `home-dir`: dir to store werf cache files and dirs (default $WERF_HOME or ~/.werf).

`tmp-dir` and `home-dir` directories store files created in runtime and cache such as git repositories and user mounts.

### Example
```
docker run                                     \
  -i --rm                                      \
  --privileged                                 \
  -v ~/.werf:/root/.werf                       \
  -v /tmp:/tmp                                 \
  -v /var/run/docker.sock:/var/run/docker.sock \
  docker:git sh <<'EOF'

# Download werf.
wget https://dl.bintray.com/flant/werf/v1.0.5-alpha.19/werf-linux-amd64-v1.0.5-alpha.19 -O /tmp/werf
chmod +x /tmp/werf
mv /tmp/werf /usr/local/bin/werf

# Create simple werf.yaml.
cat <<'EOF2' > werf.yaml
project: werf
configVersion: 1
---
image: app
from: alpine
shell:
  install: date
EOF2

# build
werf build -s=:local
EOF
```

**Comment by alexey-igrychev** [comment]
https://werf.io/documentation/v1.2/advanced/ci_cd/run_in_container/use_docker_container.html

**Comment by ilya-lesikov** [comment]
Now the preferred way to run werf in containers is by using [Buildah backend for building images](https://werf.io/getting_started/?usage=localDev&os=linux&buildBackend=buildah), not DIND.
